File: src/main/java/com/tang/designpatterns/chap1/scala/BookShelf.scala
package com.tang.designpatterns.chap1.scala

import java.util


class BookShelf(maxSize: Int) extends Aggregate {

  val books: Array[Book] = new Array[Book](maxSize)
  var last: Int = 0

  def getBookAt(index: Int): Book = books(index)

  def appendBook(book: Book) =
    books(last) = book
    last += 1

  def getLength: Int = last

  override def iterator(): MyIterator = new BookShelfIterator(this)

  def reverseIterator() = new BookShelfReverseIterator()

  class BookShelfReverseIterator extends MyIterator :

    var index = {
      val i = books.indexOf(null) - 1
      println(i)
      i
    }

    override def hasNext(): Boolean = index >= 0

    override def next(): Any =
      val book = getBookAt(index)
      index -= 1
      book

}

File: src/main/java/com/tang/designpatterns/chap1/java/BookShelf.java
package com.tang.designpatterns.chap1.java;


import java.util.ArrayList;
import java.util.List;

/**
 * @Title: BookSelf
 * @Description:
 * @author: tangyao
 * @date: 2022/9/28 10:26
 * @Version: 1.0
 */

public class BookShelf implements Aggregate {
    private List<Book> books;
    private int last = 0;

    public BookShelf(int maxsize) {
        this.books = new ArrayList<>(maxsize);
    }

    public Book getBookAt(int index) {
        return books.get(index);
    }

    public void appendBook(Book book) {
        this.books.add(book);
        last++;
    }

    public int getLength() {
        return last;
    }

    @Override
    public Iterator iterator() {
        return new BookShelfIterator(this);
    }
}